clfmerge.1
.TH "clfmerge" "1" "0.06" "Russell Coker <russell@coker.com.au>" "logtools"
.SH "NAME"
clfmerge \- merge Common\-Log Format web logs based on time\-stamps

.SH "SYNOPSIS"
.B clfmerge [\-\-help | \-h] [\-b size] [\-d] [file names]

.SH "DESCRIPTION"
The
.B clfmerge
program is designed to avoid using sort to merge multiple web log files.  Web
logs for big sites consist of multiple files in the >100M size range from a
number of machines.  For such files it is not practical to use a program such
as gnusort to merge the files because the data is not always entirely in order
(so the merge option of gnusort doesn't work so well), but it is not in random
order (so doing a complete sort would be a waste).  Also the date field that
is being sorted on is not particularly easy to specify for gnusort (I have
seen it done but it was messy).
.P
This program is designed to simply and quickly sort multiple large log files
with no need for temporary storage space or overly large buffers in memory
(the memory footprint is generally only a few megs).

.SH "OVERVIEW"
It will take a number (from 0 to n) of file\-names on the command line, it will
open them for reading and read CLF format web log data from them all.  Lines
which don't appear to be in CLF format (NB they aren't parsed fully, only
minimal parsing to determine the date is performed) will be rejected and
displayed on standard\-error.
.P
If zero files are specified then there will be no error, it will just silently
output nothing, this is for scripts which use the
.B find
command to find log files and which can't be counted on to find any log files,
it saves doing an extra check in your shell scripts.
.P
If one file is specified then the data will be read into a 1000 line buffer
and it will be removed from the buffer (and displayed on standard output) in
date order.  This is to handle the
case of web servers which date entries on the connection time but write them
to the log at completion time and thus generate log files that aren't in
order (Netscape web server does this \- I haven't checked what other web
servers do).
.P
If more than one file is specified then a line will be read from each file,
the file that had the earliest time stamp will be read from until it returns a
time stamp later than one of the other files.  Then the file with the earlier
time stamp will be read.  With multiple files the buffer size is 1000 lines or
100 * the number of files (whichever is larger).  When the buffer becomes full
the first line will be removed and displayed on standard output.

.SH "OPTIONS"
.TP 
.B \-b buffer\-size
Specify the buffer\-size to use, if 0 is specified then it means to disable the
sliding\-window sorting of the data which improves the speed.

.TP 
.B \-d
Set domain\-name mangling to on.  This means that if a line starts with
.b www.company.com
as the name of the site that was requested then that would be removed from
the start of the line and the
.B "GET /"
would be changed to
.B "GET http://www.company.com/"
which allows programs like Webalizer to produce good graphs for large hosting
sites.  Also it will make the domain name in lower case.

.SH "EXIT STATUS"
.B 0
No errors
.P
.B 1
Bad parameters
.P
.B 2
Can't open one of the specified files
.P
.B 3
Can't write to output
.SH "AUTHOR"
This program, its manual page, and the Debian package were written by
Russell Coker <russell@coker.com.au>.

.SH "SEE ALSO"
.BR clfsplit (1), clfdomainsplit (1)


debian/changelog
logtools (0.13d) unstable; urgency=low

  * Merged changes from NMU and fixed some more GCC 4.3 warnings.

 -- Russell Coker <russell@coker.com.au>  Fri, 06 Jun 2008 15:34:35 +1000

logtools (0.13c+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with GCC 4.3 (Closes: #417376).

 -- Luk Claes <luk@debian.org>  Sun, 09 Mar 2008 17:03:20 +0000

logtools (0.13c) unstable; urgency=low

  * Made it compile with GCC 3.2.

 -- Russell Coker <russell@coker.com.au>  Fri, 25 Apr 2003 17:09:00 +1000

logtools (0.13b) unstable; urgency=low

  * Improved documentation that matches how things work in sarge.

 -- Russell Coker <russell@coker.com.au>  Sun,  1 Sep 2002 15:43:00 +0200

logtools (0.13a) unstable; urgency=LOW

  * Made clfdomainsplit support URLs with the port number.

 -- Russell Coker <russell@coker.com.au>  Sun,  2 Dec 2001 23:58:00 +0100

logtools (0.13) unstable; urgency=LOW

  * Made it work with GCC 3.x.
    Closes: #121284

 -- Russell Coker <russell@coker.com.au>  Mon, 26 Nov 2001 21:46:00 +0100

logtools (0.12b) unstable; urgency=LOW

  * Fixed some minor documentation bugs, mainly related to funnel.
    Closes: #112883

 -- Russell Coker <russell@coker.com.au>  Thu, 20 Sep 2001 15:40:02 +0200

logtools (0.12a) unstable; urgency=LOW

  * Fixed header file errors on funnel for hppa.
    Closes: #104730

 -- Russell Coker <russell@coker.com.au>  Sat, 14 Jul 2001 20:04:55 +0200

logtools (0.12) unstable; urgency=LOW

  * clfmerge -d now converts domain names to lower case.

  * clfdomainsplit now converts all domain names to lower case internally
    (for sorting and matching purposes), it still exposes upper-case domains
    on its output.

 -- Russell Coker <russell@coker.com.au>  Wed, 11 Jul 2001 17:09:00 +0200

logtools (0.11) unstable; urgency=LOW

  * Made the "-d" option of clfmerge handle a URL that doesn't contain any '/'
    characters.

 -- Russell Coker <russell@coker.com.au>  Tue, 12 Jun 2001 11:58:49 +0200

logtools (0.10) unstable; urgency=MEDIUM

  * Made clfdomainsplit handle IP addresses and send them to the default file
    (first code to actually use the default file).

  * Fixed clfmerge so that the -d option will not mess with a URL that is
    already fully qualified (checks for "http://" at the start).

 -- Russell Coker <russell@coker.com.au>  Thu, 7 Jun 2001 23:49:55 +0200

logtools (0.08) unstable; urgency=MEDIUM

  * Made clfdomainsplit easier to hack (the base code could be used for
    splitting any text file if 3 small functions are changed).

  * Make clfdomainsplit use multiple processes to solve file handle limits,
    this code isn't yet tested...

 -- Russell Coker <russell@coker.com.au>  Mon, 8 May 2001 13:50:29 +0200

logtools (0.07) unstable; urgency=MEDIUM

  * Updated the description with more content and fixed some formatting issues.
    Closes: #78976

  * Made it clean the config.cache file (and other config data) before
    packaging the source.
    Closes: #96290

  * Fixed a bug where clfmerge would output the last line twice if there was
    only a single log file and buffering was on.

  * Now "domain name mangling" actually works for clfmerge!

  * Added new clfdomainsplit program.

 -- Russell Coker <russell@coker.com.au>  Mon, 7 May 2001 13:52:41 +0200

logtools (0.06) unstable; urgency=low

  * Add option to change buffer size for clfmerge.

  * Also added an option for "domain name mangling" to clfmerge.  This means
    that the web logs are expected to have a domain name preceeding the regular
    CLF format (as suggested in the Apache documentation for mass virtual
    hosting).
    See http://www.apache.org/DELETE_ME/vhosts/mass.html .

  * Fixed a bunch of errors in the man pages.

 -- Russell Coker <russell@coker.com.au>  Wed, 2 May 2001 18:11:24 +0200

logtools (0.05) unstable; urgency=low

  * Fixed many trivial bugs.

 -- Russell Coker <russell@coker.com.au>  Tue, 5 Dec 2000 21:28:54 +0100

logtools (0.04) unstable; urgency=low

  * Fixed heaps of bugs and ported to Solaris.

 -- Russell Coker <russell@coker.com.au>  Tue, 14 Nov 2000 21:39:00 +0100

logtools (0.01) unstable; urgency=low

  * Initial Release.

 -- Russell Coker <russell@coker.com.au>  Fri, 25 Aug 2000 19:53:08 +0200



clfmerge.cpp
#include <stdio.h>
#include <ext/hash_map>
#include <stdlib.h>
#include <map>
#include <cstring>
#include <unistd.h>
#include <ctype.h>

#include "logtools.h"

#define BUF_SIZE 4096

using namespace std;
using namespace __gnu_cxx;

struct eqstr
{
  bool operator()(const char *s1, const char *s2) const
  { return strcmp(s1, s2) == 0; }
};

hash_map<const char *, const char *, hash<const char *>, eqstr> months;

class LogFile
{
public:
  LogFile();
  ~LogFile(){ if(m_fp) fclose(m_fp); }

  // open a file, return 0 for success, 1 for minor error (0 length file)
  // and 2 for serious error (file not found)
  int open(const char *name, bool domain_mangling);

  // get the date string of the current item
  const char *date() const { return m_date; }
  // get the full data for the current item
  const char *line() { m_valid = false; return m_line; }

  // get a line from the file and parse it
  // return 0 for success and 1 for EOF
  int getLine();

  bool valid() const { return m_valid; }

private:
  bool m_valid;
  FILE *m_fp;
  char m_date[17];
  char m_lineBuf[BUF_SIZE];
  char m_lineBuf2[BUF_SIZE + 7];
  char *m_line;
  bool m_web_first;

  // store the date in numeric format so that strcmp() can be used to compare
  // dates.  Returns 1 for error and 0 for OK.
  int setDate();

  LogFile(const LogFile&);
  LogFile & operator=(const LogFile&);
};

LogFile::LogFile()
 : m_valid(false)
 , m_fp(NULL)
{
  m_line = m_lineBuf;
  m_lineBuf[0] = '\0';
}

// Common log format:
// 1.2.3.4 - - [23/Aug/2000:12:00:32 +0200] etc
int LogFile::setDate()
{
  unsigned int i;
  // find '[' or abort if we can't
  for(i = 0; m_lineBuf[i] != '[' && m_lineBuf[i] != '\0'; i++)
  { }

  // if not enough data left for the full date then return
  if(i + 21 > strlen(m_lineBuf))
    return 1;
  memcpy(m_date, &m_lineBuf[i + 8], 4);
  char mon[4];
  memcpy(mon, &m_lineBuf[i + 4], 3);
  mon[3] = '\0';
  const char *m = months[mon];
  if(!m) return 1;
  strcpy(&m_date[4], m);
  memcpy(&m_date[6], &m_lineBuf[i + 1], 2);
  memcpy(&m_date[8], &m_lineBuf[i + 13], 8);
  m_date[16] = '\0';
  if(m_web_first) // make m_lineBuf2 have the data for the mangled line
  {
    unsigned int end_webname;
    // find where the domain name ends
    for(end_webname = 0; m_lineBuf[end_webname] != ' ' && m_lineBuf[end_webname] != '\0'; end_webname++)
    { }

    if(end_webname >= i)
      return 1;
    for(i = 0; i < end_webname; i++)
      m_lineBuf[end_webname] = tolower(m_lineBuf[end_webname]);

    // there will be more than 40 chars in between
    unsigned int start_url = end_webname + 40;
    // search for the start quote character
    for(; m_lineBuf[start_url] != '\"' && m_lineBuf[start_url] != '\0'; start_url++)
    { }
    // search for the space in the web request
    for(; m_lineBuf[start_url] != ' ' && m_lineBuf[start_url] != '\0'; start_url++)
    { }

    if(strlen(&m_lineBuf[start_url]) < 6) return 1;

    memcpy(m_lineBuf2, &m_lineBuf[end_webname + 1], start_url - end_webname);
    m_line = &m_lineBuf2[start_url - end_webname]; // m_line points to next char
    if(strncmp(&m_lineBuf[start_url + 1], "http://", 7))
    {
      strcpy(m_line, "http://");
      m_line += 7;
      memcpy(m_line, m_lineBuf, end_webname);
      m_line += end_webname;
      if(m_lineBuf[start_url + 1] != '/')
      {
        // if URL doesn't start with a '/' then we add one
        *m_line = '/';
        m_line++;
      }
    }
    strcpy(m_line, &m_lineBuf[start_url + 1]);
    m_line = m_lineBuf2;
  }
  return 0;
}

int LogFile::open(const char *name, bool domain_mangling)
{
  m_web_first = domain_mangling;
  m_fp = fopen(name, "r");
  if(!m_fp)
  {
    fprintf(stderr, "Can't open %s.\n", name);
    return 2;
  }
  if(getLine())
    return 1;
  return 0;
}

int LogFile::getLine()
{
  while(1)
  {
    // if can't get more data then return 1
    if(!fgets(m_lineBuf, sizeof(m_lineBuf) - 1, m_fp))
      return 1;
    m_lineBuf[sizeof(m_lineBuf) - 1] = '\0';
    m_line = m_lineBuf;
    strtok(m_line, "\n\r");
    // if setDate() returns 1 then we can't parse the line so we keep looping
    // if setDate() returns 0 then return success!
    if(!setDate())
    {
      m_valid = true;
      return 0;
    }
    fprintf(stderr, "Skipping bad line: %s\n", m_line);
  }
  return 0; // to make compilers happy - will not be reached
}

typedef LogFile *PLogFile;

int item_compare(const void *a, const void *b)
{
  const LogFile * const left = *(LogFile * const *)a;
  const LogFile * const right = *(LogFile * const *)b;
  return strcmp(left->date(), right->date());
}

struct ltstr
{
  bool operator()(const string s1, const string s2) const
  {
    return strcmp(s1.c_str(), s2.c_str()) < 0;
  }
};

void usage(const char *const arg)
{
  fprintf(stderr, "usage: %s [filenames]", arg);
  fprintf(stderr, "\n"
  "This program merges web logs in common log format into a single stream\n"
  "on standard output.  It reads from multiple input files and outputs the\n"
  "data in-order as much as is possible.  If there is only a single input\n"
  "file it will re-order it (with a 1000 line buffer size) to deal with web\n"
  "servers that output data out of order.\n"
  "\nVersion: " VERSION "\n");
  exit(ERR_PARAM);
}

int main(int argc, char **argv)
{
  if(argc == 1)
    return 0;

  unsigned int map_items = 0;
  bool set_map_items = false, domain_mangling = false;
  int int_c;
  optind = 0;
  while(-1 != (int_c = getopt(argc, argv, "b:hd")) )
  {
    switch(char(int_c))
    {
      case '?':
      case ':':
      case 'h':
        usage(argv[0]);
      break;
      case 'b':
        set_map_items = true;
        map_items = atoi(optarg);
      break;
      case 'd':
        domain_mangling = true;
      break;
    }
  }
  months["Jan"] = "01";
  months["Feb"] = "02";
  months["Mar"] = "03";
  months["Apr"] = "04";
  months["May"] = "05";
  months["Jun"] = "06";
  months["Jul"] = "07";
  months["Aug"] = "08";
  months["Sep"] = "09";
  months["Oct"] = "10";
  months["Nov"] = "11";
  months["Dec"] = "12";

  multimap<const string, const string, ltstr> outputMap;

  LogFile **items = new PLogFile[argc - optind];

  unsigned int item_count = 0;
  int i;
  for(i = optind; i < argc; i++)
  {
    items[item_count] = new LogFile;
    int rc = items[item_count]->open(argv[i], domain_mangling);
    // if rc==2 then file not found, if rc==1 then 0 length file
    if(rc > 1)
      return ERR_INPUT;
    if(rc == 1)
      delete items[item_count];
    else
      item_count++;
  }

  if(!set_map_items)
  {
    map_items = item_count * 400;
    if(map_items < 4000)
      map_items = 4000;
  }
  while(item_count > 1)
  {
    qsort(items, item_count, sizeof(LogFile *), item_compare);
    while(items[0]->valid() && strcmp(items[0]->date(), items[1]->date()) <= 0)
    {
      if(map_items > 0)
      {
        outputMap.insert(pair<string, string>(items[0]->date(), items[0]->line()));
        while(outputMap.size() > map_items)
        {
          printf("%s\n", outputMap.begin()->second.c_str());
          outputMap.erase(outputMap.begin());
        }
      }
      else
      {
        printf("%s\n", items[0]->line());
      }
      if(items[0]->getLine())
      {
        delete(items[0]);
        item_count--;
        items[0] = items[item_count];
        break;
      }
    }
  }
  if(item_count == 1)
  {
    if(map_items > 0)
    {
      do
      {
        outputMap.insert(pair<string, string>(items[0]->date(), items[0]->line()));
      } while(!items[0]->getLine() && outputMap.size() < map_items);

      if(items[0]->valid())
      {
        do
        {
          outputMap.insert(pair<string, string>(items[0]->date(), items[0]->line()));
          CPCCHAR tmp = outputMap.begin()->second.c_str();
          if(printf("%s\n", tmp) != int(strlen(tmp) + 1))
          {
            fprintf(stderr, "Can't write output!\n");
              return ERR_OUTPUT;
          }
          outputMap.erase(outputMap.begin());
        } while(!items[0]->getLine());
      }
      delete items[0];
      while(!outputMap.empty())
      {
        CPCCHAR tmp = outputMap.begin()->second.c_str();
        if(printf("%s\n", tmp) != int(strlen(tmp) + 1))
        {
          fprintf(stderr, "Can't write output!\n");
            return ERR_OUTPUT;
        }
        outputMap.erase(outputMap.begin());
      }
    }
    else
    {
      do
      {
        CPCCHAR tmp = items[0]->line();
        if(printf("%s\n", tmp) != int(strlen(tmp) + 1))
        {
          fprintf(stderr, "Can't write output!\n");
            return ERR_OUTPUT;
        }
      } while(!items[0]->getLine());
    }
  }
  delete items;
  return 0;
}
